services/src/main/java/org/keycloak/authentication/authenticators/browser/RecoveryAuthnCodesFormAuthenticator.java

```diff
@@ -0,0 +1,159 @@
+
+import org.keycloak.authentication.AuthenticationFlowContext;
+import org.keycloak.authentication.AuthenticationFlowError;
+import org.keycloak.authentication.Authenticator;
+import org.keycloak.authentication.authenticators.util.AuthenticatorUtils;
+import org.keycloak.common.util.ObjectUtil;
+import org.keycloak.credential.CredentialModel;
+import org.keycloak.events.Details;
+import org.keycloak.events.Errors;
+import org.keycloak.forms.login.LoginFormsProvider;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserCredentialManager;
+import org.keycloak.models.UserCredentialModel;
+import org.keycloak.models.UserModel;
+import org.keycloak.models.credential.RecoveryAuthnCodesCredentialModel;
+import org.keycloak.models.utils.RecoveryAuthnCodesUtils;
+import org.keycloak.models.utils.FormMessage;
+import org.keycloak.services.messages.Messages;
+import org.keycloak.sessions.AuthenticationSessionModel;
+
+import javax.ws.rs.core.MultivaluedMap;
+import javax.ws.rs.core.Response;
+import java.util.Optional;
+
+import static org.keycloak.services.validation.Validation.FIELD_USERNAME;
+
+public class RecoveryAuthnCodesFormAuthenticator implements Authenticator {
+
+    private final UserCredentialManager userCredentialManager;
+
+    public RecoveryAuthnCodesFormAuthenticator(KeycloakSession keycloakSession) {
+        this.userCredentialManager = keycloakSession.userCredentialManager();
+    }
+
+    @Override
+    public void authenticate(AuthenticationFlowContext context) {
+        context.challenge(createLoginForm(context, false, null, null));
+    }
+
+    @Override
+    public void action(AuthenticationFlowContext context) {
+        context.getEvent().detail(Details.CREDENTIAL_TYPE, RecoveryAuthnCodesCredentialModel.TYPE);
+        if (isRecoveryAuthnCodeInputValid(context)) {
+            context.success();
+        }
+    }
+
+    private boolean isRecoveryAuthnCodeInputValid(AuthenticationFlowContext authnFlowContext) {
+        boolean result = false;
+        MultivaluedMap<String, String> formParamsMap;
+        String recoveryAuthnCodeUserInput;
+        RealmModel targetRealm;
+        UserModel authenticatedUser;
+        Response responseChallenge;
+        boolean isValid;
+        Optional<CredentialModel> optUserCredentialFound;
+        RecoveryAuthnCodesCredentialModel recoveryCodeCredentialModel = null;
+
+        formParamsMap = authnFlowContext.getHttpRequest().getDecodedFormParameters();
+        recoveryAuthnCodeUserInput = formParamsMap.getFirst(RecoveryAuthnCodesUtils.FIELD_RECOVERY_CODE_IN_BROWSER_FLOW);
+
+        if (ObjectUtil.isBlank(recoveryAuthnCodeUserInput)) {
+            authnFlowContext.forceChallenge(createLoginForm(authnFlowContext, true,
+                    RecoveryAuthnCodesUtils.RECOVERY_AUTHN_CODES_INPUT_DEFAULT_ERROR_MESSAGE,
+                    RecoveryAuthnCodesUtils.FIELD_RECOVERY_CODE_IN_BROWSER_FLOW));
```

Comment:
early exit here could help to reduce nesting and make the code easier to follow.

Reply:
OK.